### PR TITLE
Add rrelu

### PR DESCRIPTION
**Pull request by Bright86**
Related #437
I am new to open source but have some interest in ML. I want to do something with addons and this seems to be a good chance for me to enhance my skills.

### CODE REVIEW

**Comment by Bright86** [comment]
@googlebot I signed it!

**Comment by Bright86** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here with `@googlebot I signed it!` and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Ftensorflow%2Faddons%2Fpull%2F542) for more info**.

@googlebot I signed it!

**Comment by Bright86** [comment]
PTAL. @seanpmorgan @facaiy

**Comment by fsx950223** [comment]
> Hi @Bright86. Welcome to addons community 
> I do not think this implementation is correct. [Empirical Evaluation of Rectified Activations in Convolution Network](https://arxiv.org/pdf/1505.00853.pdf) states
> 
> 1. rrelu has different behavior in training and testing phase. So it would be better to pass `training` flag into it. See equation (5) in the paper for more details.
> 2. The random factor alpha is applied element-wise, meaning that if input size is `[10, 5]`, then RNG will generate 50 alphas in total (one alpha for one element) whenever we invoke the operation. See equation (4) in the paper for more details.
> 
> Besides, I prefer C++/CUDA OpKernel to pure Python operations for activation functions due to performance issue. I think @fsx950223 might have some thoughts on RReLU. Thanks!

I agree with you. I implement a kernel version before. But I don't know how to cache alpha which is generated in the kernel. @WindQAQ

**Comment by WindQAQ** [comment]
> I agree with you. I implement a kernel version before. But I don't know how to cache alpha which is generated in the kernel and need to be used in the gradient operation. @WindQAQ

You could make the op have two outputs. One is rrelu and the other is alpha. Note that it means you might have to inherit from `OpKernel` instead of `UnaryElementWiseOp`.

**Comment by fsx950223** [comment]
> > I agree with you. I implement a kernel version before. But I don't know how to cache alpha which is generated in the kernel and need to be used in the gradient operation. @WindQAQ
> 
> You could make the op have two outputs. One is rrelu and the other is alpha. Note that it means you might have to inherit from `OpKernel` instead of `UnaryElementWiseOp`.

Some reference?

**Comment by fsx950223** [comment]
Could someone give me a hand?
``` cpp
    float a[6];
    typename tensorflow::TTypes<float>::ConstTensor features(a,2,3);

    features.setConstant(1.0f);
    auto b = features + features.random();
```
```
bazel-out/k8-opt/bin/external/local_config_tf/include/unsupported/Eigen/CXX11/src/Tensor/TensorRandom.h:202:23: error: uninitialized const 'values' [-fpermissive]
     EIGEN_ALIGN_MAX T values[packetSize];
                       ^~~~~~
bazel-out/k8-opt/bin/external/local_config_tf/include/unsupported/Eigen/CXX11/src/Tensor/TensorRandom.h:212:17: error: assignment of read-only location 'values[j]'
       values[j] = RandomToTypeUniform<T>(&m_state, i);
```
It shows above compile errors when I use tensorflow::TTypes<float>::ConstTensor but it works well with Eigen::Tensor<float,2>.
``` cpp
    Eigen::Tensor<float,2> features(2,3);

    features.setConstant(1.0f);
    auto b = features + features.random();
```
How can I fix the compile bug?

**Comment by WindQAQ** [comment]
> Could someone give me a hand?
> 
> ```c++
>     float a[6];
>     typename tensorflow::TTypes<float>::ConstTensor features(a,2,3);
> 
>     features.setConstant(1.0f);
>     auto b = features + features.random();
> ```
> 
> ```
> bazel-out/k8-opt/bin/external/local_config_tf/include/unsupported/Eigen/CXX11/src/Tensor/TensorRandom.h:202:23: error: uninitialized const 'values' [-fpermissive]
>      EIGEN_ALIGN_MAX T values[packetSize];
>                        ^~~~~~
> bazel-out/k8-opt/bin/external/local_config_tf/include/unsupported/Eigen/CXX11/src/Tensor/TensorRandom.h:212:17: error: assignment of read-only location 'values[j]'
>        values[j] = RandomToTypeUniform<T>(&m_state, i);
> ```
> 
> It shows above compile errors when I use tensorflow::TTypes::ConstTensor::random but it works well with Eigen::Tensor<float,2>::random.
> 
> ```c++
>     Eigen::Tensor<float,2> features(2,3);
> 
>     features.setConstant(1.0f);
>     auto b = features + features.random();
> ```
> 
> How can I fix the compile bug?

It's because `features` is a ConstTensor.

**Comment by WindQAQ** [comment]
I suppose the function signature should be something like this. Don't forget to make `alpha ~ U(lower, upper)`.
```cpp
void operator()(const Device& d, typename TTypes<T>::ConstTensor features,
               T lower, T upper, bool training, typename TTypes<T>::Tensor alpha,
               typename TTypes<T>::Tensor activations) {
    alpha.device(d) = alpha.random();
    // TODO
}
```

**Comment by fsx950223** [comment]
I have a finished version sent to you. You can see it in gmail.

Tzu-Wei Sung <notifications@github.com<mailto:notifications@github.com>> 于2019年10月5日周六 下午1:54写道：

I suppose the function signature should be something like this. Don't forget to make alpha ~ U(lower, upper).

void operator()(const Device& d, typename TTypes<T>::ConstTensor features,
               T lower, T upper, bool training, typename TTypes<T>::Tensor alpha,
               typename TTypes<T>::Tensor activations) {
    alpha.device(d) = alpha.random();
    // TODO
}

**Comment by fsx950223** [comment]
> ```c++
> alpha.random();
> ```

Fixed. Thank you for your help

**Comment by WindQAQ** [comment]
Close due to inactivity. Move forward to #573. Thanks!
